src/Microsoft.Extensions.Logging.Abstractions/LoggerMessage.cs

```diff
@@ -245,11 +245,9 @@ private static LogValuesFormatter CreateLogValuesFormatter(string formatString,
             return logValuesFormatter;
         }
 
-        private class LogValues : IReadOnlyList<KeyValuePair<string, object>>
+        private readonly struct LogValues : IReadOnlyList<KeyValuePair<string, object>>
         {
-            public static Func<object, Exception, string> Callback = (state, exception) => ((LogValues)state)._formatter.Format(((LogValues)state).ToArray());
-
-            private static object[] _valueArray = new object[0];
+            public static readonly Func<LogValues, Exception, string> Callback = (state, exception) => state.ToString();
```

Comment:
Use delegate and `in` rather than `Func`?
```csharp
public delegate string Formatter(in LogValues state, Exception exception);
public static readonly Formatter Callback = (in LogValues state, Exception exception) => state.ToString();
```